src/schema.jl

```diff
@@ -237,6 +239,150 @@ function apply_schema(t::ConstantTerm, schema::Schema, Mod::Type)
+
+julia> modelmatrix(f.rhs, d)
+4×2 Array{Float64,2}:
+ 1.0  1.48861
+ 1.0  2.21097
+ 1.0  3.95192
+ 1.0  4.9999
+
+julia> d.a .+ d.b
+4-element Array{Float64,1}:
+ 1.4886128300795012
+ 2.210968202158536
+ 3.951916339835734
+ 4.999904658898614
+```
+"""
+protect(ctx) = Protected{ctx}()
+# construct singletons to avoid method ambiguities using Type{<:Protected}
+
+function apply_schema(t::FunctionTerm, schema::Schema, Mod::Type)
+    args = apply_schema.(t.args, schema, protect(Mod))
+    FunctionTerm(t.f, args, t.exorig)
+end
+
+apply_schema(t::FunctionTerm, schema::Schema, Ctx::Protected) =
+    FunctionTerm(t.f, apply_schema.(t.args, schema, Ctx), t.exorig)
+apply_schema(t, schema::Schema, Ctx::Protected) = t
+
+function apply_schema(t::FunctionTerm{typeof(protect)}, schema::Schema, Ctx::Type)
+    tt = only(t.args)
+    apply_schema(tt, schema, protect(Ctx))
+end
+
+# protect in Protected context is a no-op
+function apply_schema(t::FunctionTerm{typeof(protect)}, schema::Schema, Ctx::Protected)
+    tt = only(t.args)
+    apply_schema(tt, schema, Ctx)
+end
+
+"""
+    unprotect(term)
+    unprotect(::Protected{T})
+
+Inside a [`@formula`], removes [`Protected`](@ref) status for the argument
+term(s).  This allows the usual special [`@formula`](@ref) interpretation of
+calls to `+`, `&`, `*`, and `~` to be restored inside an otherwise
+[`Protected`](@ref) context.
+
+When called outside a `@formula`, unwraps `Protected{T}` to `T`.
+
+# Example
+
+```jldoctest; setup = :(using Random; Random.seed!(1))
+julia> d = (y=rand(4), a=[1.:4;], b=rand(4));
+
+julia> f = @formula(y ~ 1 - unprotect(a&b));
+
+julia> modelmatrix(f, d)
+4×1 Array{Float64,2}:
+  0.5113871699204988
+  0.5780635956829281
+ -1.855749019507202
+ -2.9996186355944543
+
+julia> 1 .- d.a .* d.b
+4×1 Array{Float64,2}:
+  0.5113871699204988
+  0.5780635956829281
+ -1.855749019507202
+ -2.9996186355944543
+```
+"""
+unprotect(::Protected{Ctx}) where Ctx = Ctx
+
+function apply_schema(t::FunctionTerm{typeof(unprotect)}, schema::Schema, Ctx::Protected)
+    tt = only(t.args)
+    apply_schema(tt, schema, unprotect(Ctx))
+end
+
+# un-protecting special syntax: &, +, and *
+# 
+# if these occur as children of a non-special call, they are protected by
+# default.  so if we encounter them during apply_schema and we're NOT
+# in a Protected context, we need to call the corresponding op on the
+# # arguments of the FunctionCall
+for op in (+, &, *)
+    @eval begin
+        apply_schema(t::FunctionTerm{typeof($op)}, sch::Schema, Mod::Type) =
+            apply_schema(t.f(t.args...), sch, Mod)
+    end
+end
+
+# alternatively, could define an "unprotect" macro if it's usefule elsewhere too
+# macro unprotect(op)
+#     esc(quote
+#         apply_schema(t::StatsModels.FunctionTerm{typeof($op)}, sch::StatsModels.Schema, Mod::Type) =
+#             apply_schema(t.f(t.args...), sch, Mod)
+#     end)
+# end
+
+# @unprotect(+)
+# @unprotect(&)
+# @unprotect(*)
```

Comment:
As you prefer, but either use it or remove it?

Reply:
okay, having thought about this a bit more, I think this _is_ useful since it essentially implements the recommended pattern for custom syntax now (to make methods for the function that operate on terms and return some kind of `AbstractTerm` that we can operate on with `apply_schema`, `modlecols`, etc.).